ARROW-5803: [CI] Dockerize C++ with clang 7 Travis CI

fsaintjacques opened the pull request:
Adds a new docker-compose entry (cpp-ci) which uses system and bundled packages.

- This creates an indirect dependency on the docker-hub ursalab/arrow:cpp-ci images that is pulled. The downside of this, is that any change to the Dockerfile itself will _not_ be reflected in the travis run until someone pushes upstream. This is also a problem with the `java-all-jdks` image.
- The coredump reporting functionality will not work with this image, once we migrate (if possible) all cpp based (and derived) images, we'll put this back.
- We could improve the runtime if we could build the bundled dependencies at image build step, instead of in the cmake run.

wesm commented:
> This is also a problem with the java-all-jdks image.

I think this is more of a feature than a problem. We should want to cache these images

fsaintjacques commented:
@wesm I just wanted to make it clear to reviewers the downside, e.g. discrepancy with travis and what the changeset Dockerfile has. I also think it's better for runtime. Though, we should have an automated nightly build of docker images (+ push).

fsaintjacques commented on docker-compose.yml:
```diff
@@ -114,10 +114,31 @@ services:
+      ARROW_USE_ASAN: "ON"
+      ARROW_USE_UBSAN: "ON"
+    volumes:
+      - .:/arrow:delegated
```
@kszucs can you advise how to use yaml references with an extra entry? I initially tried:
```
volumes: *ubuntu-volumes
    - ${HOME}/.ccache:/build/ccache:cached

kszucs replied:
I was also failed to merge yaml lists, so we should stick with the repetitions.

kszucs commented:
The downside can be removed with `docker-compose build --pull cpp-ci`, see its help:
```
docker-compose build --help
Build or rebuild services.

Services are built once and then tagged as `project_service`,
e.g. `composetest_db`. If you change a service's `Dockerfile` or the
contents of its build directory, you can run `docker-compose build` to rebuild it.

Usage: build [options] [--build-arg key=val...] [SERVICE...]

Options:
    --compress              Compress the build context using gzip.
    --force-rm              Always remove intermediate containers.
    --no-cache              Do not use cache when building the image.
    --pull                  Always attempt to pull a newer version of the image.
    -m, --memory MEM        Sets memory limit for the build container.
    --build-arg key=val     Set build-time variables for services.
    --parallel              Build images in parallel.
```

This the image will be built every time, but if the layers are the same then it will be as fast as pulling the image.

kszucs commented on .travis.yml:
```diff
@@ -62,28 +53,13 @@ matrix:
-    # If either C++ or Python changed, we must install the C++ libraries
     - git submodule update --init
-    - $TRAVIS_BUILD_DIR/ci/travis_before_script_cpp.sh
+    - docker-compose pull cpp-ci
```
```suggestion
    - docker-compose build --pull cpp-ci
```

kszucs commented on cpp/Dockerfile.ubuntu-bionic:
```diff
@@ -17,78 +17,102 @@
+ARG LLVM_APT_ARCH="llvm-toolchain-bionic-${LLVM_VERSION}"
+RUN wget -q -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
+RUN apt-add-repository -y --update "deb ${LLVM_APT_URL} ${LLVM_APT_ARCH} main"
+RUN apt-get install -y -q --no-install-recommends \
```
Shrink these three steps into a single one.

fsaintjacques replied:
The goal is to make the build phase much faster when one doesn't change LLVM_VERSION, the packages are huge and slow to download. I think we should keep it as-is.

fsaintjacques replied:
Note, I did this because I toyed often with removing/adding packages in the list while refatoring.

wesm replied:
I agree with combining the three RUN steps into one

fsaintjacques replied:
Will do.

kszucs commented on docker-compose.yml:
```diff
@@ -114,10 +114,31 @@ services:
-      PARQUET_TEST_DATA: /arrow/cpp/submodules/parquet-testing/data
     volumes: *ubuntu-volumes
 
+  cpp-ci:
```
I'd use a more descriptive name which highlights that we're building against system dependencies.

wesm commented:
One big picture question around the use of Docker images is how to have PRs that update the image without disturbing builds on master. Thoughts?

fsaintjacques commented:
I believe that @kszucs proposed solution (`docker-compose build --pull`) fixes this. The downside is that until the image is built and pushed upstream, the build phase is always going to be ran instead of downloading the image.

bkietz commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -2303,7 +2303,7 @@ macro(build_orc)
     endif()
     if("${COMPILER_VERSION}" VERSION_GREATER "4.0")
       set(ORC_CMAKE_CXX_FLAGS " -Wno-zero-as-null-pointer-constant \
--Wno-inconsistent-missing-destructor-override ")
```
What breaks if this is warning is absent? Could you add a TODO so this can be removed when the issue is resolved in orc?

fsaintjacques replied:
I don't think we should be tracking upstream dependencies warning.

wesm commented on docker-compose.yml:
```diff
@@ -114,10 +114,31 @@ services:
+      args:
+        LLVM_VERSION: 7
+    environment:
+      ARROW_FLIGHT: "OFF"
```
Is there a reason to turn this off?

fsaintjacques replied:
Because it was originally off in the travis build I migrated. I initially turned it off to speed up (the develop cycle with travis is excruciatingly slow) the build.

I can turn it back on again, it doesn't cut much build time, the bulk of the time is spent building protobuf & grpc, which is not triggered by this flag (but the protobuf_Source=BUNDLED...).

On this subject, we might investigate how to make grpc stop complaining about php and ruby versions. This would allow us to use the system packaged one.

fsaintjacques commented:
Rebased and updated regarding comments. Note that I did not push the image to docker-hub, the goal is to test the `docker-compose build --pull`. Once this is green, I'll push.

fsaintjacques commented:
@kszucs ready for a last review

wesm commented:
This increases Travis CI build times for this entry by about 50%, do you know why that is?

fsaintjacques commented:
@wesm because we need to build some dependencies from sources (grpc, protobuf, thrift, cares, flatbuffers), protobuf being the longest. If we disable flight, we can use ubuntu's protobuf package.

wesm commented:
Can you open a JIRA about adding some pre-built toolchain dependencies to the image, to improve runtimes (similar to how we cache thirdparty dependencies in the manylinux1 images -- I suggest we reuse the same scripts)?
